barber/base.py

```diff
@@ -0,0 +1,111 @@
+from .parameters import PartitionParametrization
+from tomo_challenge.metrics import compute_snr_score
+from scipy.optimize import minimize
+import numpy as np
+
+
+# This object evaluates a proposed splitting into a pre-defined number of bins.
+# It first maps nbin-1 parameters to the partition using a Splitter
+# It makes a Decision Tree to fit to that splitting, and then applies it to training data,
+# and then runs the challenge metric
+class EdgeSpecifierMethod:
+    def __init__(self, nbin,
+                 training_data, training_z, 
+                 validation_data, validation_z, quiet=False):
+        self.nbin = nbin
+        self.quiet = quiet
+        # Data sets
+        self.training_data = training_data
+        self.training_z = training_z
+        self.validation_data = validation_data
+        self.validation_z = validation_z
+
+        self.partitioner = PartitionParametrization(nbin)
+        self.zmin = training_z.min()
+        self.zmax = training_z.max()
+        self.zrange = self.zmax - self.zmin
+
+
+    def __call__(self, p):
+        # Split into partition parameters and any extra parameters
+        r = p[self.nbin - 1:]
+        p = p[:self.nbin - 1]
+
+        z_edges = self.unit_parameters_to_z_edges(p)
+        score = self.evaluate_edges(z_edges, r)
+        return score
+    
+    def unit_parameters_to_z_edges(self, p):
+        n1 = len(p)
+        n2 = self.nbin - 1
+        if not n1 == n2:
+            raise ValueError("Wrong size (should be {n2} but is {n1})")
+        # Convert to a partition of the unit interval
+        q = self.partitioner(p)
+        # And convert that to a partition of zmin .. zmax
+        z_edges = np.concatenate([[self.zmin], self.zmin + self.zrange * np.cumsum(q)])
+        return z_edges
+
+    def evaluate_edges(self, bin_edges, extra_parameters):
+        if not self.quiet:
+            print(f"Edges: {bin_edges}   Extra: {extra_parameters}")
+
+        # Assign each galaxy in the training data to a bin
+        training_bin = np.digitize(self.training_z, bin_edges, right=True) - 1
```

Comment:
I'm not sure I understand what `base.py` is doing -- are we not defining "bin edges" solely in the space of photometric observations, without the training/test set distinction?